### PR TITLE
Autodeps + shakeCommandOptions

### PR DESCRIPTION
**Pull request by jacereda**

### CODE REVIEW

File: src/Development/Shake/CmdOption.hs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    | Traced String -- ^ Name to use with 'traced', or @\"\"@ for no tracing. By default traces using the name of the executable.
+    | Timeout Double -- ^ Abort the computation after N seconds, will raise a failure exit code. Calls 'interruptProcessGroupOf' and 'terminateProcess', but may sometimes fail to abort the process and not timeout.
+    | WithStdout Bool -- ^ Should I include the @stdout@ in the exception if the command fails? Defaults to 'False'.
+    | WithStderr Bool -- ^ Should I include the @stderr@ in the exception if the command fails? Defaults to 'True'.
+    | EchoStdout Bool -- ^ Should I echo the @stdout@? Defaults to 'True' unless a 'Stdout' result is required or you use 'FileStdout'.
+    | EchoStderr Bool -- ^ Should I echo the @stderr@? Defaults to 'True' unless a 'Stderr' result is required or you use 'FileStderr'.
+    | FileStdout FilePath -- ^ Should I put the @stdout@ to a file.
+    | FileStderr FilePath -- ^ Should I put the @stderr@ to a file.
+    | Autodeps -- ^ Compute dependencies automatically. Requires either 'LintFSATrace' or 'LintTracker'.
+      deriving (Eq,Ord,Show,Data)
```
Comment (ndmitchell) [inline src/Development/Shake/CmdOption.hs]: You'll need to add Typeable to the Deriving for pre-7.10

File: src/Development/Shake/Command.hs [hunk 2]
```diff
@@ -154,6 +134,8 @@ commandExplicit funcName copts results exe args = do
                             , any (`isPrefixOf` x) inside
                             , not $ any ($ x) ignore]
                 rel = map (makeRelative cwd)
+            when (Autodeps `elem` copts) $
+              needed $ rel $ ham rs
```
Comment (ndmitchell) [inline src/Development/Shake/Command.hs]: 4 space indents are preferred, although not really a big deal (but I would reindent it if I'm ever near that line).

File: src/Development/Shake/Types.hs [hunk 3]
```diff
@@ -162,27 +165,28 @@ data ShakeOptions = ShakeOptions
...
-    gfoldl k z (ShakeOptions x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11 x12 x13 x14 x15 x16 x17 x18 x19 x20 y1 y2) =
+    gfoldl k z (ShakeOptions x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11 x12 x13 x14 x15 x16 x17 x18 x19 x20 x21 y1 y2) =
         z unhide `k` x1 `k` x2 `k` x3 `k` x4 `k` x5 `k` x6 `k` x7 `k` x8 `k` x9 `k` x10 `k` x11 `k`
-        x12 `k` x13 `k` x14 `k` x15 `k` x16 `k` x17 `k` x18 `k` x19 `k` x20 `k`
+        x12 `k` x13 `k` x14 `k` x15 `k` x16 `k` x17 `k` x18 `k` x19 `k` x20 `k` x21 `k`
         Function y1 `k` Function y2
-    gunfold k z c = k $ k $ k $ k $ k $ k $ k $ k $ k $ k $ k $ k $ k $ k $ k $ k $ k $ k $ k $ k $ k $ k $ z unhide
+    gunfold k z c = k $ k $ k $ k $ k $ k $ k $ k $ k $ k $ k $ k $ k $ k $ k $ k $ k $ k $ k $ k $ k $ k $ k $ z unhide
     toConstr ShakeOptions{} = conShakeOptions
     dataTypeOf _ = tyShakeOptions
```
Comment (ndmitchell) [inline src/Development/Shake/Types.hs]: I'm strongly considering removing the Data instance. The original idea was to help with people writing cmdargs command line option handling, but I suspect everyone uses the built in one instead, and the options are a bit too complex for that. Do you have any thoughts? (Not a comment on this pull request, just more general.)

**Comment by ndmitchell** [comment]
Looks good, modulo the deriving typeable (and anything else the buildbots spot). Just off to bed, but I'll merge tomorrow, assuming it's all passing then.

**Comment by ndmitchell** [comment]
BTW, in general, ignore the Appveyor failures - I'm still working on bringing them up to spec. The Travis failures. In this case, you need to add the new module to the .cabal file as well.

File: src/Development/Shake/Types.hs (hunk 3 above)
Comment (jacereda) [inline src/Development/Shake/Types.hs]: Why isn't it derived automagically?

Comment (jacereda) [inline src/Development/Shake/Types.hs]: Not that I use it, I wouldn't care if it was removed...

File: src/Development/Shake/Command.hs (hunk 2 above)
Comment (jacereda) [inline src/Development/Shake/Command.hs]: Damn, I added a .dir-locals.el just for that, but it's on another machine :-(
Maybe we could add one to the repo with your preferred settings?

File: src/Development/Shake/Types.hs (hunk 3 above)
Comment (ndmitchell) [inline src/Development/Shake/Types.hs]: You can't derive Data for types containing functions. If it was auto derived I wouldn't care as much.

**Comment by ndmitchell** [comment]
Merged, thanks! I also renamed Autodeps to AutoDeps, which seems nicer (splits the words properly). Let me know if that seems like a bad idea.

**Pull request merged**